### ISSUE TITLE
Need advice on Solid Queue's memory usage

### ISSUE THREAD

**Issue by yjchieng** [rails/solid_queue#330]
Ruby: 3.3.4
Rails: 7.2.1
Solid Queue: 0.7.0, 0.8.2

--------------

I run a Rails App on AWS EC2 instance with 1G of memory.
I notice the solid queue process takes up 15-20% of the instance's memory, which becomes the single largest process by memory usage.

--------------

What I checked:

1) Check memory usage by start/stop supervisorctl 
(I use it to manage my solid queue process)

stop supervisorctl - free memory 276MB
start supervisorctl - free memory 117MB
> It increases 159MB

2) Stop supervisorctl service, and run "solid_queue:start" directly
Trying to see if this is something related to supervisor.

before solid_queue:start - free memory 252MB
after solid_queue:start - free memory 109MB
> It increases 143MB

3) Then I notice there is a latest version.
I upgraded to 0.8.2 (was 0.7.0).

stop supervisorctl - free memory 220MB
start supervisorctl - free memory 38MB
> It increases 182MB

--------------

I need some advise:

1) Is 150-200MB the minimum requirement to run "solid_queue:start"?
2) Is there any setting/feature that I can switch off to reduce memory usage?
3) Is there any setting that I can limit the maximum memory usage?

And, thanks a lot for making this wonderful gem. :)

**Comment by rosa** [comment]
Hey @yjchieng, thanks for opening this issue! 🙏 I think it depends a lot on your app. A brand new Rails app seems to use around 74.6MB memory for me after booting (without Solid Queue, just running Puma). I think the consumption you're seeing is from all the processes together and not just the supervisor, measuring free memory before starting the supervisor and after, as the supervisor will fork more processes. Are you running multiple workers or just one? I think reducing the number of workers there would help. Another thing that might help is using `bin/jobs`, which preloads the whole app before forking, but the gains there are usually quite modest.

**Comment by rosa** [comment]
There might also be something else going on because the only changes from version 0.7.0 to 0.8.2 were for the installing part of Solid Queue; nothing was changed besides the initial installation, so the memory footprint shouldn't have changed. I imagine there is other stuff going on in your AWS instance at the same time that might consuming memory as well.

**Comment by Focus-me34** [comment]
Up 🆙🔥 
I have huge memory issues in production (Rails 7.2 + activeJob + solidQueue). Everything works just fine in dev mode, but in production, there seems to be a memory leak. After restarting my production server, I get to roughly ~75% RAM usage. Very quickly (talking in minutes...) I get to ~100%. And if I let the app run for the weekend and come back on Monday (like today), I get to... 288% RAM usage... I tried removing all the lines in my code related to solidQueue, and I can confirm that this is what's causing the memory issue in production.

Any advice/suggestions would be very much appreciated fellow devs. Have an amazing day!

**Comment by rosa** [comment]
@Focus-me34, what version of Solid Queue are you running? And when you say you're removing anything related to Solid Queue, what Active Job adapter are you using instead?

**Comment by Focus-me34** [comment]
@rosa I'm using Solid Queue version 1.0.0. I checked all sub-dependencies versions. They all respect the pre-requirements.
We didn't really try with any other adapter since Solid Queue will be the default adapter in RoR 8. Hence, we really want to make it work this way.

Here's some of our setup code:

```
# scrape_rss_feed_job.rb
class ScrapingJob < ApplicationJob
  queue_as :default
  limits_concurrency to: 1, key: -> { "rss_feed_job" }, duration: 1.minute

  def perform
    Api::V1::EntriesController.fetch_latest_entries
  end
end
```

```
# recurring.yml
default: &default
  periodic_cleanup:
    class: ScrapeSecRssFeedJob
    schedule: every 2 minutes

development:
  <<: *default

test:
  <<: *default

production:
  <<: *default
```

```
# queue.yml
default: &default
  dispatchers:
    - polling_interval: 1
      batch_size: 500
      concurrency_maintenance_intervaL: 15
  workers:
    - queues: "*"
      threads: 3
      processes: <%= ENV.fetch("JOB_CONCURRENCY", 1) %>
      polling_interval: 0.1

development:
  <<: *default

test:
  <<: *default

production:
  <<: *default
```

Do you see anything weird?

**Comment by rosa** [comment]
No, that looks good to me re: configuration. You said: 

> I tried removing all the lines in my code related to solidQueue, and I can confirm that this is what's causing the memory issue in production.

So, if you don't get any memory issues when not using Solid Queue, is that because you're not running any jobs at all, if you're not using another adapter? Because that would point to the jobs having a memory leak, not Solid Queue.

**Comment by Focus-me34** [comment]
Hey Rosa, sorry for the delayed reply! I've been very busy at work.

Here’s where we're at: I've been working on getting our company’s code running smoothly with Rails 7.2 and Solid Queue (in production Heroku). As I mentioned earlier, it’s been a huge challenge, and unfortunately, we haven’t had much success with it.

My colleague and I decided to take a closer look at our code to see if the problem was on our end. Since my last comment here, we’ve implemented tests, and I can confirm that the code is behaving exactly as expected.

Our next step, after troubleshooting the high memory usage on Heroku, was to try switching away from Solid Queue and try a different job adapter (as you suggested). I've set up Sidekiq as the adapter, and we saw a drastic improvement: memory usage dropped from around 170% of our 512 MB quota to a range of 25%-70%.

This leads me to believe that there might be a memory leak in production when using Solid Queue. From our observations, it seems that after the initial job execution completes, instance variables at the top of the function (which should reset to nil at the start of each job) are retaining the values from the previous iteration. We suspect this might be preventing the Garbage Collector from clearing memory properly between jobs.

Let me know if there's any more information I can provide to help you investigate. We’re really looking forward to moving back to using the built-in Solid Queue functionality once this issue is resolved.

**Comment by rajeevriitm** [comment]
I am observing high memory usage when solid queue is used on heroku as well. Has there been any solutions to fix this issue? is there any temporary fix that can be used for now?

**Comment by arikarim** [comment]
solid queue with puma plugin is stupidly using high memory.

**Comment by rosa** [comment]
Hey @rajeevriitm, no, no real solutions. I had an async adapter that would run the supervisor, workers and dispatchers and everything together in the same process, which would save memory. However, this was scrapped from the 1.0 version. I need to push for that again. 

In the meantime, you can try using `rake solid_queue:start` to run Solid Queue, as that, by default, won't preload all your app and see if that makes a difference. What's your configuration like? 

@arikarim, thanks for your very helpful and useful comment 😒

**Comment by arikarim** [comment]
Haha sorry i was so tired of this.
until i found out that i was running rails s instead of bundle exec puma.
my bad

**Comment by arikarim** [comment]
i still there is some strange issues, with puma concurrency more than 0 the memory goes up 😢

**Comment by rajeevriitm** [comment]
@rosa I have a simple configuration. queue runs on a single server config. I have a continuous running job that runs every 30 mins.

queue.yml
```
default: &default
  dispatchers:
    - polling_interval: 2
      batch_size: 500
  workers:
    - queues: "*"
      threads: 1
      processes: 1
      polling_interval: 2

development:
  <<: *default

test:
  <<: *default

production:
  <<: *default
```

puma.rb
```
threads_count = ENV.fetch("RAILS_MAX_THREADS", 3)
threads threads_count, threads_count
port ENV.fetch("PORT", 3000)
plugin :tmp_restart

# Run the Solid Queue supervisor inside of Puma for single-server deployments
plugin :solid_queue 

pidfile ENV["PIDFILE"] if ENV["PIDFILE"]
```

**Comment by IvanPakhomov99** [comment]
+1, I'm running the latest `solid_queue` version and am definitely experiencing memory issues. Once I start the container, memory usage keeps increasing indefinitely until it hits the maximum capacity and the container restarts.

Note: I don't run any jobs.

![Image](https://github.com/user-attachments/assets/92ed5dc7-119c-4e6f-b9d0-c9d172045043)

**Comment by eliasousa** [comment]
hi 👋 I'm experiencing the same issue:

Ruby: 3.4.2
Rails: 8.0.2
Solid Queue: 1.1.4

Running on DigitalOcean droplet instance with 1G of memory. I'll try to downgrade the solid_queue gem version to see if improves

**Comment by rosa** [comment]
Hey all, so sorry about this! I've been swamped with other stuff at work, but I'm going to look into this on Monday.

**Comment by rajeevriitm** [comment]
Thanks for addressing the issue. Hope it's resolved soon.

**Comment by rosa** [comment]
@IvanPakhomov99, @rajeevriitm, could you try downgrading to version v1.1.2 and let me know if the issue persists? Also, what version of Ruby are you using?

**Comment by IvanPakhomov99** [comment]
Hi @rosa, Thanks for your prompt response. I tested four different Solid Queue versions (1.1.0, 1.1.1, 1.1.2, and 1.1.4) but encountered the same issue across all of them.

**Environment:**
- Ruby: 3.4.1
- Rails: 8.0.1
- Database: MySQL 8.0 (Solid Queue runs on the main instance)

Let me know if you need any additional details.

**Comment by rajeevriitm** [comment]
@rosa I tried downgrading. The issue exist in 1.1.2 as well. 
```
Ruby 3.4.1
Rails 8.0.1
```
Happy to help .

**Comment by rajeevriitm** [comment]
hey @rosa , were you able to identify the issue causing the memory leak?

**Comment by rosa** [comment]
@rajeevriitm, I'm afraid I wasn't 😞  I reviewed all code from v1.1.0 and didn't identify anything that could leak memory. This was before @IvanPakhomov99 shared that testing 1.1.0, 1.1.1, 1.1.2, and 1.1.4 made no difference. Then, I tried running jobs of different kinds (recurring jobs being enqueued every 5 seconds of different types, long running jobs, etc.) over a couple of days and couldn't reproduce any memory leaks. I think whatever is happening depends on what your jobs are doing or what you're loading in your app. I also wonder if this is not a memory leak, but simply high memory usage. Solid Queue runs a different process for each worker, a process for the dispatcher, another process for the scheduler and another process for the supervisor. All those processes load your app and even though the fork is done after loading the app, so CoW should ensure memory sharing, this is not the same as running a single process. 

I don't have a better idea to reduce memory usage other than providing a single-process execution mode (what I've called "async" mode).

**Comment by IvanPakhomov99** [comment]
@rosa Thanks for taking the time to look into this! The main pitfall here is that it's a brand new service and I am not running any jobs yet. I also confirmed that the service pod memory is stable.

That said, I actually have an update — I downgraded to version 1.0.2, and memory usage looks stable now. I’ll stick with this version for now.

**Comment by chayuto** [comment]
have the same issue on heroku to. with version 1.1.4. 

as mentioned above. downgraded to version 1.0.2 solve the issue.

<img width="1250" alt="Image" src="https://github.com/user-attachments/assets/61975767-5ae5-467e-a205-1ebda77b3db2" />

**Comment by jeffcoh23** [comment]
@rosa Thank you for your attentiveness to this issue.  I just downgraded `solid_queue` to 1.0.2 and can confirm that it drastically reduced the R14 related errors/memory usage.

**Comment by rosa** [comment]
Oh! Thanks a lot for confirming that! I had only looked down to version 1.1.0. I'm on-call this week so a bit short on time but will try to figure out why the memory increased from that version to 1.1.0.

**Comment by IvanPakhomov99** [comment]
@rosa I’m not sure how relevant this is, but it might be worth taking another look at this commit: https://github.com/rails/solid_queue/commit/a152f2637e69675612737e7da533ae7f2d4f092e. It looks like `interruptible_sleep` was updated to use Promises.future on each call. Given how this method is used — often multiple times within a loop — it’s quite possible this could lead to increased memory consumption.

Even though the block uses `.value`, making it synchronous from the caller’s perspective, a new thread or fiber is still created under the hood for each call.

**Comment by jeffcoh23** [comment]
> [@rosa](https://github.com/rosa) Thank you for your attentiveness to this issue. I just downgraded `solid_queue` to 1.0.2 and can confirm that it drastically reduced the R14 related errors/memory usage.

Just to follow up on this - it seemed to have worked in the short-term, but I am back to where I started unfortunately... still seeing those r14/memory issues consistently.  It is causing my jobs to not finish

**Comment by zdennis** [comment]
We are experiencing the same unbounded memory growth issues with resolve in OOMs. We're on solid-queue 1.1.3 and Ruby 3.3.7.

We can recreate this issue by running `bin/jobs` and letting it run, without enqueuing any jobs. Here's what the memory size of the solid-queue worker process `"solid-queue-worker(1.1.3): waiting for jobs in *`  looks like.

![Image](https://github.com/user-attachments/assets/3e1c2847-429c-4004-acfb-1a00f04d88ff)

It's being sampled every minute for about 75 minutes and grows from about 147mb to 840mb in that time. This was run in development with YJIT on and eager loading off.  We have ran it with YJIT off and eager loading off, and the same issue persists. 

The same memory growth issue exists in staging and production environments as well which results in eventual OOM errors. The memory growth appears unbounded as it will eventually consume all memory even if we bump it to a ridiculous amount. It never seems to plateau. 

Note: there are no jobs being enqueued during this benchmark. This is just the worker process looking for jobs.

I'm currently running with solid-queue 1.1.4 and so far the issue appears exist there as well.

**Comment by rosa** [comment]
Alright, I'm going to revert the change in https://github.com/rails/solid_queue/pull/417 since that's looking like the most likely culprit as @IvanPakhomov99 said. I hope that fixes this problem.

**Comment by zdennis** [comment]
Thanks, @rosa . 🙇  Let us know when/how to test, and I'm happy to pull it down and re-run the same tests.

**Comment by rosa** [comment]
Thanks a lot @zdennis, really appreciate it 🙏  I've just pushed version 1.1.5 with that change reverted, I hope this is it 🤞

**Comment by zdennis** [comment]
solid_queue 1.1.5 is looking much better. After about 80 minutes, memory isn't growing. 🥳 

### Worker memory 🟢 

![Image](https://github.com/user-attachments/assets/f37c95a3-fa2c-4d5d-a2a0-1ce395c6eb30)

Plateaus right around 149Mb for us.

### Supervisor memory 🟢 

And the supervisor also plateaus, right around 159Mb:

![Image](https://github.com/user-attachments/assets/fa21de11-ac9d-48c7-9764-e4bb88fb696e)

### Concurrent::Promises.future issue 

I wonder if this is related the memory leak reported in usage of Concurrent::Promises.future  in https://github.com/ruby-concurrency/concurrent-ruby/issues/960.

### Thanks for pushing out 1.1.5 🙇 

Thanks for quickly push out 1.1.5, @rosa! It is looking like it may be the ticket here. I'll keep this test running over night (because I'm curious), and will post back with those results as well.

**Comment by zdennis** [comment]
Here is the memory usage on the worker and supervisor processes over 15 hours. Again, this is with no jobs being processed. 

Still looks great compared to prior releases. We'll bump and test this out in a live environment. I'll share results later this week or next week after it's had some time running in the wild.

### Worker memory 

Worker memory is still very low at 152Mb.

![Image](https://github.com/user-attachments/assets/7f938ac3-c3a1-41ec-b9e0-ac687f44f2d8)

### Supervisor memory 

Supervisor memory is up to about 165.7mb. It very slowly grew from about 156mb to 160mb and then jumped up to 165mb after about 14 hours. 

![Image](https://github.com/user-attachments/assets/a58dd7e8-a0f3-48c1-ae68-cab8087d0701)

I haven't been benchmarking/profiling GC major/minor runs and we're not manually invoking GC.start so it's not clear if this is normal growth by Ruby or if there is a small leak in the supervisor process still. Either way, 1.1.5 continues to look good.

**Comment by rafael-pissardo** [comment]
@zdennis any supprise until now?

**Comment by zdennis** [comment]
We've had it running in production since late Monday and we haven't had an OOM since. Here's a chart showing how over a similar timeframe (from Saturday to Monday) we did see an OOM with solid_queue 1.1.3. And then over the same number of days, Monday evening to today, we haven't seen an OOM and memory looks much better.

![Image](https://github.com/user-attachments/assets/d4593a69-0d4b-40a9-b7eb-c3016782b03a)

We did notice that memory usage grows when polling (and when there are no active jobs). We have some recurring jobs and then we have past jobs.  We are not invoking `GC.start` manually so I suspect the memory growth we're seeing with polling may be a result of Ruby not garbage collecting for a while, but this is just a guess.

I notice solid_queue has a usage of `Concurrent::Promises.future_on`. I'm not sure if this is susceptible to the same issue that `Concurrent::Promises.future` is. 

I'll post back again early next week after it's been up for a week, but so far it is still looking good. 🤞

**Comment by hms** [comment]
> [@rosa](https://github.com/rosa) I’m not sure how relevant this is, but it might be worth taking another look at this commit: [a152f26](https://github.com/rails/solid_queue/commit/a152f2637e69675612737e7da533ae7f2d4f092e). It looks like `interruptible_sleep` was updated to use `Promises.future` on each call. Given how this method is used — often multiple times within a loop — it’s quite possible this could lead to increased memory consumption.
> 
> Even though the block uses `.value`, making it synchronous from the caller’s perspective, a new thread or fiber is still created under the hood for each call.
> 
> Sorry if that’s not the case — I might not be seeing the full picture. Just reviewing the changes between versions 1.1.0 and 1.0.2.

I've instrumented the code between the original (and now current) implementation of Interruptible.rb Vs. the change I made in [a152f26] and they result in exactly the same number calls to Poll.  This strongly suggests the implementation of Interruptible is not changing SQ behavior for number of iterations within the poll loop semantics. 

I've built a test harness to allow for isolating various configurations (Ruby versions, Rails versions, Database configurations, and SolidQueue configurations).  

I've tested various implementations of Concurrent::Promises.future including the Interruptible implementation in the PR that I submitted and can also confirm that it does not leak. 

While testing without any jobs (just worker polling) what I'm seeing is a lot of garbage being generated in-between major GC's, which can be exacerbated via a shorter polling_interval, and results constant and significant memory growth.  This is forcing Ruby to allocate more slabs to handle the object allocations. While the objects are GC'able, the slabs are forever (stackable via GC.stats and GC.heap_stats).

When I switch to JEMALLOC and MALLOC_MAX_ARENA=2, which is how my Heroku and local dev is configured, I don't see the memory growth issues at all.

More digging to forthcoming.

**Comment by rafael-pissardo** [comment]
Also confirming here that the leak has been resolved in v1.1.5. 👍

![Image](https://github.com/user-attachments/assets/0b0de784-f136-4d89-984e-ce3d1f852d8a)

**Comment by jrodden1** [comment]
@zdennis Any updates on how its been running in `prod` for you?  

@rosa, I hate to rain on the parade, but I'm still having the memory growth issue even after upgrading to `1.1.5` on heroku.  😕

![Image](https://github.com/user-attachments/assets/b1b7ebbf-0701-465c-8d10-98999c3bf99c)
**Context**
- Rails 8.0.2
- Ruby 3.4.2
- Solid Queue 1.1.5
  - Solid Queue is configured for a single database setup.  
- I'm running 1 web dyno for puma and 1 worker dyno for solid_queue on heroku. 
- I'm using the `jemalloc` build pack on heroku with `JEMALLOC_ENABLED` set to `true` 
- I've tried using `bundle exec bin/jobs` as well as `bundle exec rails solid_queue:start` in my Procfile.  It was just this afternoon when I switched to trying `rails solid_queue:start` in the Procfile so I'll report on how that goes... but it doesn't look promising as the memory usage is still climbing in a similar fashion. 
 
**Thoughts**
I'm testing out this setup as a `staging` environment on heroku for now, but I've got about 2 weeks until I need to deploy this to `prod` for my client. 

I have 5 recurring jobs specified in my `recurring.yml`.  These are pretty much the only jobs running (except for a mailer that runs once per day that gets triggered by of one of the jobs).  Most of the time, the queue is empty / idle but the memory usage continually grows.  I'm curious if I'm having the issue and others are not is because I have multiple recurring jobs (and one that runs once per minute)? 🤔  

I've included the content of my config files.  Perhaps I have something incorrectly configured?  I'm open to thoughts or suggestions. 🙏  Should I try adding a recurring job that forces garbage collection to run or something like that as a workaround?  

**Solid Queue Config**
```yml
# config/queue.yml
default: &default
  dispatchers:
    - polling_interval: 1
      batch_size: 500
  workers:
    - queues: "*"
      threads: 2
      processes: <%= ENV.fetch("JOB_CONCURRENCY", 1) %>
      polling_interval: 1

development:
  <<: *default
  database: app_development

test:
  <<: *default
  database: app_test

production:
  <<: *default
  url: <%= ENV["DATABASE_URL"] %>
```

**SQ Recurring Jobs Config**
(I've generic-ized the names of the jobs as this project is for a client)
```yml
# config/recurring.yml
recurring_maintenance: &recurring_maintenance
  job_one:
    class: "JobOne"
    priority: 0
    schedule: "every 60 seconds"
  job_two:
    class: "JobTwo"
    priority: 1
    schedule: "every 5 minutes"
  job_three:
    class: "JobThree"
    priority: 1
    schedule: "every 30 minutes"
  job_four:
    class: "JobFour"
    priority: 1
    schedule: "every weekday at 9am America/Detroit"
  job_five:
    class: "JobFive"
    priority: 1
    schedule: "30 15 15 * * 1-5 America/Detroit"

development:
  <<: *recurring_maintenance

production:
  <<: *recurring_maintenance
```

**Comment by zdennis** [comment]
@jrodden1 wrote:

> @zdennis Any updates on how its been running in prod for you?

Our staging and production environments are still looking good for us. 

![Image](https://github.com/user-attachments/assets/9e461d58-004c-417e-a858-d87796cb61c7)

This isn't to say there's not another issue going on. Memory growth still occurs, but it's no longer unbounded and it gets reclaimed. We have more memory available in our clusters than it appears you have in your Heroku app Heroku so it could be that if we had a lower limit, we might be seeing some OOM issues.

I am eagerly waiting @hms 's continued investigation as I think he may be on to something. I'd love to dive in deeper, but I don't have bandwidth right now.

Have you tried setting `MALLOC_MAX_ARENA=2` as @hms [mentioned](https://github.com/rails/solid_queue/issues/330#issuecomment-2828956188)?

**Comment by hms** [comment]
I have to admin I reprioritized my personal work priorities given it was assumed my patch was the root cause and since my code was removed, I moved on.

I can resume my digging, but I think I need some help.  

There are too many permutations between ruby versions, rails versions, and databases (and database versions).  The community  needs to identify a small subset of configurations to be our test beds as it's a ton of work for me to manage all of the versions that have been mentioned in this thread.

Secondly, I need better info around if people are seeing the leak with SQ simply idling, running standard jobs, and/or running recurring jobs. That's also a lot of permutations to manage.  Even worse, idling and standard jobs run on the Worker and reassuring jobs run on the Scheduler, making monitoring harder.

What I've found so far is as follows:

- I was never able to reproduce the leak with 1.1.4 when letting SQ idle (suggested above as observable).  If it was leaking it small enough that I didn't see it over my monitoring window.
- I didn't see any real differences (other than what would be expected due to the slight differences in the code) in memory usage or object counts between V1.1.4 and V1.1.5 during my idle testing.
- The Supervisor memory footprint growing over time is confusing and damning . The Supervisor doesn't really do much after SQ has launched other than heartbeating via a separate thread and waiting for, and cleaning up, broken pipes from sub-workers terminating (and as side note, wasn't changed in the code changes in question).


One test that I didn't run yet, but has me curious is around Interruptible sleep. In the 1.1.5 release, Rosa kept (by accident???) one change that I made that removed *a lot* of unnecessary worker polling. If the problem is in some manner tied to the poll loop, then by putting back the original high volume poll loop, it should make the problem a little easier to spot.

For those of you suffering from OOM issues, here is what I'm doing these days and has been working well for months:

- I have all of my jobs that generates OOM pressure restricted to a single queue with single-threaded workers.

- Using a perform_around block, I check memory before and after the job execution (the before info just helps monitor the jump for specific job types and isn't strictly needed).

-  If the Worker is over my memory limit (technically, it's the SQ in aggregate memory footprint, but I only restart the heavy memory worker)

   - execute Process.kill :SIGTERM, $$; sleep 0.2 -- which in theory allows the existing job to complete the required AJ and SQ house keeping to complete

   - prevents new jobs from starting since SIGTERM is a graceful shutdown and has terminated the poll loop..  Note: there is a small      race here between the signal being processed and actually setting the shutdown flag which prevents another job from starting.  This is the reason for the sleep -- the signal is processed asynchrony and the short sleep allows for the signal to propagate.  The sleep prevents the job from completing, which has the side effect of preventing a new job from starting.
  
   - causes the worker in question to terminate -- releasing all memory used by that worker
  
   - allows the Supervisor to restart it (hopefully without any pending jobs which take a long time before they run)

This approach has some pros and cons:

- Pro: I don't get spurious H14 errors anymore. For me, this is big.
- Con: It adds some overheads and delays for the impacted jobs. For me, this is small.
- Con: It doesn't play nice with YJIT since we're frequently restarting this worker. For me this is small and with a small PR, could be managed better.

Despite the above, it works well enough that I have abandoned a PR to implement Worker restarts as native feature of SQ.  

It should be noted that this works much more cleanly with a dedicated queue / single-threaded worker and known jobs that generate memory pressure.  If you kill a worker with multiple threads, there is the risk killing an active job and with the current SQ design, that job can take a while to be rerun or even worse, can get caught in a run, die loop.

**Comment by jrodden1** [comment]
Huzzah!  **Success!**
  
I turned my recurring jobs back on toward the end of the work day yesterday and also changed my `threads` setting to `1` in `queue.yml`.

![Image](https://github.com/user-attachments/assets/be5669e3-0384-460b-9a62-4e427b9ebc77)

With this setup, this is the memory graph I'm getting now.  

![Image](https://github.com/user-attachments/assets/123894c2-aac7-480b-a307-94d79094d7d2)

While I don't have a lot of 'headroom' (only around ~100MB), I'm happy to see that the usage leveled off. 

I'll be curious to see if this changes if/when I add additional recurring jobs, but for now, I think I'm good! 

Thanks for the input @hms, @zdennis!

**Comment by ka8725** [comment]
Having the issue with too RAM used on Heroku too. Solid queue 1.2.1. Rails 8. Ruby 3.4.6.

These are my results:

1. Changing threads to 1 didn't help me much;
2. Turning off recurring tasks helped a lot - ~2x less RAM;
3. Starting with the rake task also helps to reduce some memory, but not so much as 2.

Decided to clean jobs via scheduler. Just added this command to run every 10 minutes:

```
rails runner "SolidQueue::Job.clear_finished_in_batches(sleep_between_batches: 0.3)"
```

Not much time passed. But it looks good for now:

<img width="2576" height="988" alt="Image" src="https://github.com/user-attachments/assets/460212d2-0441-4648-a80d-09bdc4f00a1c" />

Some time after, I turned on recurring jobs and got memory leak again:

<img width="2500" height="592" alt="Image" src="https://github.com/user-attachments/assets/700b93c5-19b7-4ba1-a951-af5f7eaae4f4" />

UPD. Confirming turning off recurring jobs make RAM stable:

<img width="1250" height="296" alt="Image" src="https://github.com/user-attachments/assets/526e537c-ce16-4780-81c5-647168aea0bd" />

**Comment by onurozer** [comment]
I'm also looking for advice on Solid Queue's memory usage. It feels unusually high at 450MB+ for a basic, low traffic Rails 8 app with a single recurring job.

Setup
```
Ruby 3.4.5
Rails 8.0.2.1
Solid Queue: 1.2.1
Server: Ubuntu 24.04 (LTS) x64
```

Memory Usage:
```
supervisor: 74MB
dispatcher: 83MB
worker: 159MB
scheduler: 141MB
```

queue.yml
```
default: &default
  dispatchers:
    - polling_interval: 1
      batch_size: 500
  workers:
    - queues: "*"
      threads: 3
      processes: <%= ENV.fetch("JOB_CONCURRENCY", 1) %>
      polling_interval: 0.1
```

Solid queue runs as a separate process using `bin/jobs`. Restarting the solid queue process doesn't seem to change the memory usage.

Is the memory usage normal? Any ideas on how to reduce it?

Thanks!

**Comment by hms** [comment]
I've probably dug into managing the SolidQueue memory usage as much as anyone and all I can say at this point in time is:
* SolidQueue seems to reflect the new Rails perspective that we users of small VSP's (not just Heroku, any flavor of small deployments) need to be rethinking our deployment strategies.
  - SolidQueue made a couple of highly impactful design decisions to support the flexibility that the very largest Rails shops require, but they push the minimum deployment requirements up by 1GB+.
* There is a minimum price per process that's painfully large (even with SQ's CoW friendly design choices) because they are all using a significant amount of the Rails infrastructure.  My experience suggests that 450MB for a quiescent deployment with one instance of each of the base processes is about as low a memory footprint as you can safely assume.
* The best solution that I've found to-date is periodically restarting the non-supervisor processes. The way I do this is:
    - Only use 1 thread for Workers that I intend to restart. Otherwise it's too hard to shutdown on job completion boundaries and one risks having a job that gets killed in-flight and requires a manual restart coupled with the need to design jobs to be fully restartable (this requires more careful job design than I think most people realize, and makes all of the corner conditions come out to play).
    - Note: The dispatcher and Scheduler are never so busy that they can't complete whatever activities they are doing and then execute an orderly shutdown if you have your shutdown interval larger than 20 seconds.
    - I use an around_perform in application_job that:
      * checks memory at the end of the job (but before the around_perform block exits) and if over my threshold for either the Worker for System as a whole
       * sends SIGTERM to the processes that I want to restart.  
       * Sleep for 1 second to allow the signal to propagate
       * Then lets the job finish so SQ can do its house keeping (which it will complete before responding to the request to terminate).         
       * Since I never restart the Supervisor, it will restart any missing processes
  
GC.compact crashes with my Gems (one of them was compact unfriendly), so be careful if you choose to experiment.  I played with a SQ PR that would allow triggering workers to not start new jobs and terminate when all threads go quiescent, but that feels like a bandaid over consciously made design choices.

And for the truth in advertising part of this post, I lost the battle and was forced to upgrade my Heroku job instance (more than doubling my annual billing). I was unable to keep it within Heroku's default budget of 0.5GB, even allowing for periods of a flood of R14 errors (which if one is not careful, can burn your log storage providers logging limits) while an offending job ran, no matter how aggressive I got with triggering process restarts.

**Comment by yjchieng** [comment]
My journey is similar. 
I had increased the instance size to 1 GB RAM to allow several of my smaller applications to run comfortably with solid queue.

No significant impact for applications already operating at scale though.

**Comment by rosa** [comment]
@hms is spot on! His write-up is excellent. 

I'm so sorry, y'all. I'm very aware of this problem, and I'd like to rescue the async mode that was yanked before launch (there's even a PR for this: https://github.com/rails/solid_queue/pull/644 but it needs more work. I've been completely swamped with other work and just haven't had the time or attention available for this, but just so you know it's very much in my radar and something I'm determined to improve.

**Comment by hms** [comment]
@rosa Thank you.

One observation:  I would guess that Dispatcher and Scheduler could live together in a single process very comfortably without any impact on flexibility or functionality.  They are both designed around running periodically, with low impact if because of threading issues, one were to run slightly behind schedule (1 or 2 seconds max). For those running larger configurations where those seconds matter, it would simply be a matter of starting the process twice with different config files.

For those of us who are trying to live on a memory budget, that buys back somewhere between 100 to 150MB and that is significant -- maybe enough of a slim down to fit back into our 512Mb wardrobe.... ;-)

If you are open to this change, I would be willing to see if it's within my capabilities.

Hal

PS. What's the ruby version support timelines such that Ractors could be considered instead of threads?  At least *on paper*, they would make the merging of functionality into a single process more performant than a pure thread implementation (lots and lots of benchmarking in someone's future).

**Comment by saiqulhaq** [comment]
Rails is intentionally designed with developer experience as its top priority, operating on the principle that hardware costs are relatively low compared to developer productivity and time. This philosophy means that resource efficiency, including RAM usage, takes a back seat to enabling rapid prototyping and maintaining developer-friendly workflows.

**Comment by rosa** [comment]
Sorry @hms, missed your last comment! 

You're right that the dispatcher and the scheduler would perfectly live together in the same process. In fact, this was the case in the beginning, but I changed it to get a bit more flexibility, back when I still thought I'd have the async mode eventually. I don't like so much the idea of having that configurable, though. I think, if we were to merge just those two together, they'd be just one process for everyone. Perhaps that'd be a way and the next priority, yes 🤔 

About the Ruby version support timelines, I'm not sure! I haven't really thought about it 😳 I guess I'll follow Rails there. I think version 2.0 should drop support for Rails < 8, and follow Rails's ruby requirement, and then for Ractors... I guess they could be used in an experimental, optional way 🤔 But I haven't thought about it 😳

**Comment by hms** [comment]
@rosa 

I was asking about Ractors only to help decide if it's worth my time to experiment.  I'm not sure if they are ready for SolidQueue's requirements in 3.5 (at the very least, the N:M scheduler has to be functional and not have any surprises, and I think #port will at least put in a cameo) but it would be fun to find out.

Given the support for older Ruby's, if Ractors were interesting, would SolidQueue even consider a configurable Ractor option?

@hms

**Comment by rosa** [comment]
Hey folks! If you're still using Solid Queue and running into those problems, you could give a try to https://github.com/rails/solid_queue/pull/644. I'm hoping to include this in the next release. It reinstates the async mode that was removed long ago. It'd be nice to get some more testing done before I ship it.

**Comment by calvinyj** [comment]
> Hey folks! If you're still using Solid Queue and running into those problems, you could give a try to [#644](https://github.com/rails/solid_queue/pull/644). I'm hoping to include this in the next release. It reinstates the async mode that was removed long ago. It'd be nice to get some more testing done before I ship it.

I have deployed this branch to a AWS t4g.micro instance (1GB memory) for a small application.
```
Ruby: 4.0.0
Rails: 8.1
```

Running for about 12 hours. First ~8 hours without async mode and ~4 hours with async mode enabled.
So far did not observe anything unusual, except the process info below.


./bin/jobs start
We can see the supervisor is supervising the dispatcher, worker and scheduler (431785, 431789, 431793)
```
ubuntu    431740 34.9 20.9 1488340 195716 ?      Sl   02:36   0:03 solid-queue-fork-supervisor(1.2.4): supervising 431785, 431789, 431793
ubuntu    431785  3.2 18.1 1490420 169592 ?      Sl   02:36   0:00 solid-queue-dispatcher(1.2.4): dispatching every 1 seconds
ubuntu    431789  6.4 18.1 1490696 169452 ?      Sl   02:36   0:00 solid-queue-worker(1.2.4): waiting for jobs in *
ubuntu    431793  4.4 18.0 1490696 168720 ?      Sl   02:36   0:00 solid-queue-scheduler(1.2.4): scheduling some_jobs
```

./bin/jobs start --mode=async
I am not sure what are these (17552, 17560, 17568).
If I understand correctly, worker should be on 432249, same as supervisor?
```
ubuntu    432249 44.2 21.5 1485352 201572 ?      Sl   02:41   0:04 solid-queue-async-supervisor(1.2.4): supervising 17552, 17560, 17568
```

**Comment by rosa** [comment]
> I am not sure what are these (17552, 17560, 17568).

Those are the thread IDs. 

> If I understand correctly, worker should be on 432249, same as supervisor?

Yes! Same process.

**Comment by jrodden1** [comment]
Hi Folks!  So **I've just upgraded my application on my production server** to use the **new SQ 1.3.0** and its **new `async` mode**.  This has **drastically reduced memory usage to** an average of ~**260MB** on my Heroku dyno! 

The usage of this new `async` mode was quite easy to implement (see the updated [docs](https://github.com/rails/solid_queue?tab=readme-ov-file#fork-vs-async-mode)) but here is how I implemented it:  
* Update the version of Solid Queue gem in your Gemfile to `>=1.3.0` 
* Run `bundle install` 
* Update your `Procfile`'s `worker` line to tack on `--mode=async` to the end of the `bin/jobs` command.  
* Commit changes and deploy. 

I first tried this out on my test server to see how the `worker` dyno would do and it performed great.  All my recurring tasks were running as desired and memory consumption was way down.  

After letting that test server run overnight and still seeing the memory usage consistently low and jobs running normally, I decided to put it out to my production server yesterday.  

So far so good on prod! 

The graphic below shows the memory usage on my prod `worker` dyno before the upgrade deploy at around 3pm (I was on SQ 1.2.4 previously) and after the deploy with the upgrade to SQ 1.3.0 using `async` mode:
<img width="588" height="265" alt="Image" src="https://github.com/user-attachments/assets/e4169cf3-cdb1-430c-9f81-96b987e305d3" />

So if you are still experiencing memory usage issues, you should have a look into SolidQueue 1.3.0's `async` mode. :) 

Thanks @rosa and crew for all the work you put into this!  The new `async` option works great in my use case.

**Comment by jeffcoh23** [comment]
Wow thats amazing @jrodden1 and @rosa - do you think a basic dyno is sufficient now?  Would love to switch back from Fly to Heroku, but needing the 1gb RAM dyno on Heroku was too costly for beginner projects.

**Comment by onurozer** [comment]
I've been running in `async` mode for about a week now, and the memory usage went down from 450MB to 220MB. Amazing work, thank you @rosa and @joshleblanc  🙏

**Comment by rosa** [comment]
This is great to know! Just a note: I released version 1.3.1 a couple of days ago with a minor bugfix for the async mode, in case you want to update. 

Since this is working well for y'all, I'm going to close this issue. Thanks everyone for the patience!

**Comment by jrodden1** [comment]
> Wow thats amazing [@jrodden1](https://github.com/jrodden1) and [@rosa](https://github.com/rosa) - do you think a basic dyno is sufficient now? Would love to switch back from Fly to Heroku, but needing the 1gb RAM dyno on Heroku was too costly for beginner projects.

@jeffcoh23 Yep.  It should work on a Heroku Basic dyno (it is running well under the 512MB limit now).

**Comment by jtFrancisco** [comment]
@rosa In the documentation it mentions

> The recommended and default mode is fork. Only use async if you know what you're doing and have strong reasons to.

Could you explain some issues that using async mode might cause?

**Comment by rosa** [comment]
> Could you explain some issues that using async mode might cause?

You'll get a performance hit because of the GVL, and if you run everything in the same process, with Puma, your app's performance will be affected by your queue. Besides that, this mode has been significantly less battle-tested, and you might find some obscure, multi-threading bug that might be hard to debug, and I might not have time to help with it because I don't use this 🙈
